lib/vdc/BOVCollection.cpp

```diff
@@ -0,0 +1,503 @@
+        return FOUND;
+    }
+    return NOT_FOUND;
+}
+
+void BOVCollection::_findTokenValue(std::string &line) const
+{
+    std::string delimiter = ": ";
+
+    size_t      pos = 0;
+    std::string token;
+    while ((pos = line.find(delimiter)) != std::string::npos) {
+        token = line.substr(0, pos);
+        line.erase(0, pos + delimiter.length());
+    }
+}
+
+int BOVCollection::_sizeOfFormat(DC::XType type) const
+{
+    switch (type) {
+    case DC::XType::INT32: return 4;
+    case DC::XType::FLOAT: return 4;
+    case DC::XType::DOUBLE: return 8;
+    default: return -1;
+    }
+}
+
+void BOVCollection::_swapBytes(void *vptr, size_t size, size_t n) const
+{
+    unsigned char *ucptr = (unsigned char *)vptr;
+    unsigned char  uc;
+    size_t         i, j;
+
+    for (j = 0; j < n; j++) {
+        for (i = 0; i < size / 2; i++) {
+            uc = ucptr[i];
+            ucptr[i] = ucptr[size - i - 1];
+            ucptr[size - i - 1] = uc;
+        }
+        ucptr += size;
+    }
+}
+
+template<class T> int BOVCollection::ReadRegion(std::string varname, size_t ts, const std::vector<size_t> &min, const std::vector<size_t> &max, T region)
+{
+    float       time = _times[ts];
+    std::string dataFile = _dataFileMap[varname][time];
+
+    FILE *fp = fopen(dataFile.c_str(), "rb");
+    if (!fp) {
+        SetErrMsg("Invalid file: %s : %M", dataFile.c_str());
+        return -1;
+    }
+
+    int formatSize = _sizeOfFormat(_dataFormat);
+    if (formatSize < 0) {
+        SetErrMsg("Unspecified data format");
+        fclose(fp);
+        return (-1);
+    }
+
+    size_t numValues = _gridSize[0] * _gridSize[1] * _gridSize[2];
+
+    int  n = 1;
+    bool systemLittleEndian = *(char *)&n == 1 ? true : false;
+    bool dataLittleEndian = _dataEndian == "LITTLE" ? true : false;
+    bool needSwap = systemLittleEndian != dataLittleEndian ? true : false;
+
+    // Read a "pencil" of data along the X axis, one row at a time
+    size_t count = max[0] - min[0] + 1;
+    // Note: allocate buffer once and reuse for many times, so repeated allocation is avoided.
+    std::vector<unsigned char> vReadBuffer(count * formatSize);
+    unsigned char *            readBuffer = vReadBuffer.data();
+    for (size_t k = min[2]; k <= max[2]; k++) {
+        size_t zOffset = _gridSize[0] * _gridSize[1] * k;
+        for (size_t j = min[1]; j <= max[1]; j++) {
+            size_t xOffset = min[0];
+            size_t yOffset = _gridSize[0] * j;
+            size_t offset = formatSize * (xOffset + yOffset + zOffset) + _byteOffset;
+
+            fseek(fp, offset, SEEK_SET);
+            size_t rc = fread(readBuffer, formatSize, count, fp);
+
+            if (rc != count) {
+                if (ferror(fp) != 0) {
+                    MyBase::SetErrMsg("Error reading input file: %M");
+                } else {
+                    MyBase::SetErrMsg("Short read on input file: %M");
+                }
+                fclose(fp);
+                return -1;
```

Comment:
Throughout this code, sometime `return ERROR` is used, while some other times `return -1` is used. I suggest we strive for consistency within a module (i.e. BOV module).

Reply:
For readability, Stas made the suggestion for using an enum when these functions have multiple return codes (FOUND, NOT_FOUND, and ERROR), instead of 0, -1, and -2.   These values don't really apply to other functions though.

Reply:
I think you might want to consider C++ 11 enum classes which provides better safety and readability. 
https://www.geeksforgeeks.org/enum-classes-in-c-and-their-advantage-over-enum-datatype/

Reply:
Interesting read.  I have no problem with this.  Thanks, and fixed.

Reply:
Update: Scott decides to wait for input from John before making any change.

Reply:
Q to Scott, is this really fixed? I.e., is it consistent to use all integers or all enums?